### PR TITLE
Revise configuration-related docs (3.8 and earlier)

### PR DESCRIPTION
**Pull request by josh-wong**
## Description

This PR revises configuration-related documents.

> **Note**
>
> This PR targets ScalarDB 3.8 to 3.5 and incorporates the feedback in the original PR: #1082 
>
> A new PR is needed for reasons mentioned in [my comment in the previous PR](https://github.com/scalar-labs/scalardb/pull/1082#issuecomment-1737229546).

## Related issues and/or PRs

https://github.com/scalar-labs/scalardb/pull/1082

## Changes made

This PR includes the same changes that were addressed in https://github.com/scalar-labs/scalardb/pull/1082:

- Revised contents.
- Added notice blocks for noteworthy content.
- Added content tabs for databases (in `configurations.md`).
- Added Mermaid code for some diagrams (in `configurations.md`).
- Deleted `configurations-for-consensus-commit.md` since its contents were previously combined with `configurations.md`.

## Checklist

> The following is a best-effort checklist. If any items in this checklist are not applicable to this PR or are dependent on other, unmerged PRs, please still mark the checkboxes after you have read and understood each item.

- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes.
- [x] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [x] Tests (unit, integration, etc.) have been added for the changes.
- [x] My changes generate no new warnings.
- [x] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

N/A

## Release notes

N/A

### CODE REVIEW

File: docs/configurations.md [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+The following configurations are available for Cassandra:
+
+| Name                                    | Description                                                           | Default    |
+|-----------------------------------------|-----------------------------------------------------------------------|------------|
+| `scalar.db.storage`                     | `cassandra` must be specified.                                        | -          |
+| `scalar.db.contact_points`              | Comma-separated contact points.                                       |            |
+| `scalar.db.contact_port`                | Port number for all the contact points.                               |            |
+| `scalar.db.username`                    | Username to access the database.                                      |            |
+| `scalar.db.password`                    | Password to access the database.                                      |            |
+| `scalar.db.cassandra.metadata.keyspace` | Keyspace name for the namespace and table metadata used for ScalarDB. | `scalardb` |
```
Comment (josh-wong) [inline docs/configurations.md:76]: Since this PR is for `3.8` and earlier, I don't think this configuration should be in this doc. Is that correct?

File: docs/configurations.md [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+<div id="CosmosDB_for_NoSQL" class="tabcontent" markdown="1">
+
+The following configurations are available for CosmosDB for NoSQL:
+
+| Name                                 | Description                                                                                              | Default    |
+|--------------------------------------|----------------------------------------------------------------------------------------------------------|------------|
+| `scalar.db.storage`                  | `cosmos` must be specified.                                                                              | -          |
+| `scalar.db.contact_points`           | Azure Cosmos DB for NoSQL endpoint with which ScalarDB should communicate.                               |            |
+| `scalar.db.password`                 | Either a master or read-only key used to perform authentication for accessing Azure Cosmos DB for NoSQL. |            |
+| `scalar.db.cosmos.metadata.database` | Database name for the namespace and table metadata used for ScalarDB.                                    | `scalardb` |
```
Comment (josh-wong) [inline docs/configurations.md:88]: Since this PR is for `3.8` and earlier, I think this configuration should be as follows, right?

```suggestion
| `scalar.db.cosmos.table_metadata.database` | Database name for the namespace and table metadata used for ScalarDB.                                    | `scalardb` |
```

File: docs/configurations.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+| `scalar.db.password`                  | AWS secret access key used to authenticate the user interacting with AWS.                                                                                                                                                                                   |            |
+| `scalar.db.dynamo.endpoint_override`  | Amazon DynamoDB endpoint with which ScalarDB should communicate. This is primarily used for testing with a local instance instead of an AWS service.                                                                                                        |            |
+| `scalar.db.dynamo.metadata.namespace` | Namespace name for the namespace and table metadata used for ScalarDB.                                                                                                                                                                                      | `scalardb` |
```
Comment (josh-wong) [inline docs/configurations.md:102]: Since this PR is for `3.8` and earlier, I think this configuration should be as follows, right?

```suggestion
| `scalar.db.dynamo.table_metadata.namespace` | Namespace name for the namespace and table metadata used for ScalarDB.                                                                                                                                                                                      | `scalardb` |
```

File: docs/configurations.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+| `scalar.db.jdbc.prepared_statements_pool.enabled`         | Setting this property to `true` enables prepared-statement pooling.                                                                                                          | `false`                      |
+| `scalar.db.jdbc.prepared_statements_pool.max_open`        | Maximum number of open statements that can be allocated from the statement pool at the same time. Use a negative value for no limit.                                                  | `-1`                         |
+| `scalar.db.jdbc.isolation_level`                          | Isolation level for JDBC. `READ_UNCOMMITTED`, `READ_COMMITTED`, `REPEATABLE_READ`, or `SERIALIZABLE` can be specified.                                                       | Underlying-database specific |
+| `scalar.db.jdbc.metadata.schema`                          | Schema name for the namespace and table metadata used for ScalarDB.                                                                                                          | `scalardb`                   |
```
Comment (josh-wong) [inline docs/configurations.md:122]: Since this PR is for `3.8` and earlier, I think this configuration should be as follows, right?

```suggestion
| `scalar.db.jdbc.table_metadata.schema`                          | Schema name for the table metadata used for ScalarDB.                                                                                                          | `scalardb`                   |
```

File: docs/configurations.md (hunk 1 above)
Comment (Torch3333) [reply docs/configurations.md:76]: Yes exactly.

File: docs/two-phase-commit-transactions.md [hunk 5]
```diff
@@ -263,113 +347,409 @@ interface ServiceB {
...
+        commit(transaction1, transaction2);
+      } catch (UnsatisfiedConditionException e) {
+        // You need to handle `UnsatisfiedConditionException` only if a mutation operation specifies 
+        // a condition. This exception indicates the condition for the mutation operation is not met.
+
+        rollback(transaction1, transaction2);
+
+        // You can handle the exception here, according to your application requirements.
+
+        return;
```
Comment (Torch3333) [inline docs/two-phase-commit-transactions.md:528]: This exception is part of a [feature](https://github.com/scalar-labs/scalardb/pull/899) for the future ScalarDB 4.0 so we can remove it.
```suggestion

```

File: docs/two-phase-commit-transactions.md [hunk 6]
```diff
@@ -263,113 +347,409 @@ interface ServiceB {
...
+
+- If you catch `CrudException`, this exception indicates that the transaction CRUD operation has failed due to transient or non-transient faults. You can try retrying the transaction from the beginning, but the transaction will still fail if the cause is non-transient.
+- If you catch `CrudConflictException`, this exception indicates that the transaction CRUD operation has failed due to transient faults (for example, a conflict error). In this case, you can retry the transaction from the beginning.
+
+### `UnsatisfiedConditionException`
+
+The APIs for mutation operations (`put()`, `delete()`, and `mutate()`) could also throw `UnsatisfiedConditionException`.
+
+If you catch `UnsatisfiedConditionException`, this exception indicates that the condition for the mutation operation is not met. You can handle this exception according to your application requirements.
+
```
Comment (Torch3333) [inline docs/two-phase-commit-transactions.md:680]: Likewise here
```suggestion
```

File: docs/configurations.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+| `scalar.db.metadata.cache_expiration_time_secs`                  | ScalarDB has a metadata cache to reduce the number of requests to the database. This setting specifies the expiration time of the cache in seconds.                                                               | `-1` (no expiration) |
+| `scalar.db.active_transaction_management.expiration_time_millis` | ScalarDB maintains ongoing transactions, which can be resumed by using a transaction ID. This setting specifies the expiration time of this transaction management feature in milliseconds.                       | `-1` (no expiration) |
+| `scalar.db.default_namespace_name`                               | The given namespace name will be used by operations that do not already specify a namespace. If you would like to use this setting with ScalarDB Server, configure this setting in the client-side configuration. |                      |
```
Comment (Torch3333) [inline docs/configurations.md:193]: `scalar.db.default_namespace_name` do not exist for ScalarDB <= 3.8, so we can remove it.
```suggestion
```

**Review by Torch3333 — COMMENTED** [review]
Thank you for backporting the changes to these docs. It's pretty tricky to keep track of the feature getting deprecated or appearing.

File: docs/multi-storage-transactions.md [hunk 8]
```diff
@@ -1,53 +1,60 @@
...
+This page explains how multi-storage transactions work and how to configure the feature in ScalarDB.
 
-Internally, the `multi-storage` implementation holds multiple storage instances and has mappings
-from a namespace name to a proper storage instance. When an operation is executed, it chooses a
-proper storage instance from the specified namespace by using the namespace-storage mappings and
-uses it.
+## How multi-storage transactions work in ScalarDB
 
-## Configuration
+In ScalarDB, the `multi-storage` implementation holds multiple storage instances and has mappings from a namespace name to a proper storage instance. When an operation is executed, the multi-storage transactions feature chooses a proper storage instance from the specified namespace by using the namespace-storage mapping and uses that storage instance.
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/multi-storage-transactions.md:9]: Got it. When backporting this doc to version branches `3.5` and `3.6`, I will change this paragraph as follows, which includes mentions of `table name` and `table-storage`. (The additional wording matches the wording used in this doc on version branches `3.6` and `3.5`.

```suggestion
In ScalarDB, the `multi-storage` implementation holds multiple storage instances and has mapping from a table name or a namespace name to a proper storage instance. When an operation is executed, the multi-storage transactions feature chooses a proper storage instance from the specified table or namespace by using the table-storage or namespace-storage mapping and uses that storage instance.
```

File: docs/configurations.md (hunk 7 above)
Comment (josh-wong) [reply docs/configurations.md]: Got it. I've removed that part.

File: docs/configurations.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+|------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------|
+| `scalar.db.metadata.cache_expiration_time_secs`                  | ScalarDB has a metadata cache to reduce the number of requests to the database. This setting specifies the expiration time of the cache in seconds.                                                               | `-1` (no expiration) |
+| `scalar.db.active_transaction_management.expiration_time_millis` | ScalarDB maintains ongoing transactions, which can be resumed by using a transaction ID. This setting specifies the expiration time of this transaction management feature in milliseconds.                       | `-1` (no expiration) |
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/configurations.md:174]: Got it. I will remove this from version branches `3.7`, `3.6`, and `3.5` when I backport this PR.

File: docs/configurations.md [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+| Name                                              | Description                                                                                      | Default                 |
+|---------------------------------------------------|--------------------------------------------------------------------------------------------------|-------------------------|
+| `scalar.db.server.port`                           | Port number for ScalarDB Server.                                                                 | `60051`                 |
+| `scalar.db.server.prometheus_exporter_port`       | Prometheus exporter port. Prometheus exporter will not be started if a negative number is given. | `8080`                  |
+| `scalar.db.server.grpc.max_inbound_message_size`  | The maximum message size allowed to be received.                                                 | The gRPC default value. |
+| `scalar.db.server.grpc.max_inbound_metadata_size` | The maximum size of metadata allowed to be received.                                             | The gRPC default value. |
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/configurations.md:163]: Got it. I will remove this when backporting this doc to version branch `3.5`.

File: docs/configurations.md [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+| `scalar.db.server.decommissioning_duration_secs`  | The decommissioning duration in seconds.                                                         | `30`                    |
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/configurations.md]: Got it. Since this PR targets version branches `3.8` and earlier, I've removed this configuration.

File: docs/configurations.md [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+The following configurations are available for the gRPC transaction manager for ScalarDB Server:
+
+| Name                                       | Description                                                 | Default                 |
+|--------------------------------------------|-------------------------------------------------------------|-------------------------|
+| `scalar.db.transaction_manager`            | `grpc` should be specified.                                 | -                       |
+| `scalar.db.contact_points`                 | ScalarDB Server host.                                       |                         |
+| `scalar.db.contact_port`                   | Port number for ScalarDB Server.                            | `60051`                 |
+| `scalar.db.grpc.deadline_duration_millis`  | The deadline duration for gRPC connections in milliseconds. | `60000` (60 seconds)    |
+| `scalar.db.grpc.max_inbound_message_size`  | The maximum message size allowed for a single gRPC frame.   | The gRPC default value. |
+| `scalar.db.grpc.max_inbound_metadata_size` | The maximum size of metadata allowed to be received.        | The gRPC default value. |
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/configurations.md:148]: Got it. I will remove these when backporting this doc to version branch `3.5`.

File: docs/configurations.md [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+| `scalar.db.consensus_commit.coordinator.namespace`    | Namespace name of Coordinator tables.                                                                                                                                                                                                                                                                                                                                                                   | `coordinator` |
+| `scalar.db.consensus_commit.include_metadata.enabled` | If set to `true`, `Get` and `Scan` operations results will contain transaction metadata. To see the transaction metadata columns details for a given table, you can use the `DistributedTransactionAdmin.getTableMetadata()` method, which will return the table metadata augmented with the transaction metadata columns. Using this configuration can be useful to investigate transaction-related issues. | `false`       |
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/configurations.md:35]: Got it. I will remove this when backporting this doc to version branches `3.7`, `3.6`, and `3.5`.

**Review by josh-wong — COMMENTED** [review]
@Torch3333 Thank you for your comments!🙏 I've incorporated many of your comments and will incorporate the rest when backporting PRs to earlier branches.

File: docs/multi-storage-transactions.md [hunk 14]
```diff
@@ -1,53 +1,60 @@
...
-# JDBC specific configurations for the "mysql" storage. As mentioned before, the format is "scalar.db.multi_storage.storages.<storage name>.<property name without the prefix 'scalar.db.'>". So for example, if you want to specify the "scalar.db.jdbc.connection_pool.min_idle" property for the "mysql" storage, you can specify "scalar.db.multi_storage.storages.mysql.jdbc.connection_pool.min_idle"
+# Define the JDBC-specific configurations for the "mysql" storage.
 scalar.db.multi_storage.storages.mysql.jdbc.connection_pool.min_idle=5
 scalar.db.multi_storage.storages.mysql.jdbc.connection_pool.max_idle=10
 scalar.db.multi_storage.storages.mysql.jdbc.connection_pool.max_total=25
 
-# Define namespace mappings from a namespace name to a storage. The format is "<namespace name>:<storage name>,..."
+# Define namespace mapping from a namespace name to a storage.
+# The format is "<NAMESPACE_NAME>:<STORAGE_NAME>,...".
 scalar.db.multi_storage.namespace_mapping=user:cassandra,coordinator:mysql
```
[the remark this answers is not in the document]
Comment (brfrn169) [reply docs/multi-storage-transactions.md:51]: @Torch3333 Thank you! Yes, we should mention the `table mapping` configuration for ScalarDB <= 3.6.

File: docs/two-phase-commit-transactions.md [hunk 15]
```diff
@@ -143,107 +157,177 @@ tx.put(fromPut);
...
 }
 ```
 
-If an error happens, you need to call `rollback()` (or `abort()`) in all the coordinator/participant processes.
+For `prepare()`, if any of the Coordinator or participant processes fail to prepare the transaction, you will need to call `rollback()` (or `abort()`) in all the Coordinator and participant processes.
 
-You can call `prepare()`, `commit()`, `rollback()` in the coordinator/participant processes in parallel for better performance.
+For `commit()`, if any of the Coordinator or participant processes successfully commit the transaction, you can consider the transaction as committed. When a transaction has been committed, you can ignore any errors in the other Coordinator and participant processes. If all the Coordinator and participant processes fail to commit the transaction, you will need to call `rollback()` (or `abort()`) in all the Coordinator and participant processes.
+
+For better performance, you can call `prepare()`, `commit()`, and `rollback()` in the Coordinator and participant processes in parallel, respectively.
```
[the remark this answers is not in the document]
Comment (brfrn169) [reply docs/two-phase-commit-transactions.md:191]: @Torch3333 (CC: @josh-wong) Yes, that's for `3.8` and later. Thank you for mentioning that!

File: docs/two-phase-commit-transactions.md
Comment (brfrn169) [inline docs/two-phase-commit-transactions.md:1]: Maybe we should also rename this file?

Comment (josh-wong) [reply docs/two-phase-commit-transactions.md:1]: At some point, yes. However, since a link to this doc is included in many other docs in this repository and other repositories, I think we should leave it as is for now so that we don't break the link.

File: docs/configurations.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+| `scalar.db.metadata.cache_expiration_time_secs`                  | ScalarDB has a metadata cache to reduce the number of requests to the database. This setting specifies the expiration time of the cache in seconds.                                                               | `-1` (no expiration) |
+| `scalar.db.active_transaction_management.expiration_time_millis` | ScalarDB maintains ongoing transactions, which can be resumed by using a transaction ID. This setting specifies the expiration time of this transaction management feature in milliseconds.                       | `-1` (no expiration) |
+
+### Two-phase commit support
+
+ScalarDB supports transactions with a two-phase commit interface. With transactions with a two-phase commit interface, you can execute a transaction that spans multiple processes or applications, like in a microservice architecture.
+
+For details about using two-phase commit, see [Transactions with a Two-Phase Commit Interface](two-phase-commit-transactions.md).
+
+## Configuration example - App and database
```
Comment (brfrn169) [inline docs/configurations.md:182]: Why did you remove the ScalarDB Server example here? In 3.8 and earlier, we don't deprecate it yet?

File: docs/configurations.md [hunk 17]
```diff
@@ -0,0 +1,216 @@
+# ScalarDB Configurations
+
+This page describes the available configurations for ScalarDB.
+
+## Transaction managers
+
+Implemented within ScalarDB are two transaction managers: Consensus Commit and gRPC. You can specify one of the transaction managers by using the `scalar.db.transaction_manager` property and then specify the configurations for the transaction manager.
```
Comment (feeblefakie) [inline docs/configurations.md:7]: Although we can set `grpc` in `scalar.db.transaction_manager`,
I think it's pretty misleading to say we have two transaction managers because gRPC is not a transaction manager.
(More correctly, I feel Consens Commit is a transaction protocol, not a transaction manager.)

How about saying something like this? > @josh-wong @brfrn169 
(Maybe the sentences are broken up or we can reorganize the sections.)


ScalarDB provides its own transaction protocol called Consens Commit. You can use the Consensus Commit protocol directly through the ScalarDB client library or through a gRPC-based ScalarDB Server, which is a daemon process that manages ScalarDB transactions. 
When you use the Consensus Commit protocol directly through the ScalarDB client library, you specify `scalar.db.transaction_manager=consensus-commit` or you can omit the configuration since the `consensus-commit` is the default value.
When you use the Consensus Commit protocol through a gRPC-based ScalarDB Server, you need to specify `scalar.db.transaction_manager=grpc`.

Comment (brfrn169) [reply docs/configurations.md:7]: Sounds good! Thank you.

File: docs/configurations.md (hunk 16 above)
Comment (josh-wong) [reply docs/configurations.md]: Sorry, I forgot to re-add the example after realizing that we would address 3.8 first (the example was removed in the [previous PR that I closed](https://github.com/scalar-labs/scalardb/pull/1082/files/b3a9d9f0990323abef4a561d6db50c8beb931b82#r1322606138)). I've re-added the example.

File: docs/two-phase-commit-transactions.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/two-phase-commit-transactions.md:191]: Got it. I will make sure that this part is not backported to `3.7`, `3.6`, and `3.5`.

File: docs/configurations.md (hunk 17 above)
Comment (josh-wong) [reply docs/configurations.md]: @feeblefakie Thank you for the explanation!

I added a sentence about specifying `scalar.db.transaction_manager=grpc` to the `## ScalarDB Server (gRPC)` section (a sentence about specifying `scalar.db.transaction_manager=consensus-commit` already exists in the `## Consensus Commit` section), so I've simplified this paragraph with the first couple of sentences that you provided.

Since you feel that Consensus Commit is a transaction protocol rather than a transaction manager, what do you think about changing the heading for this section from `## Transaction managers` to `## Transaction protocols`? If so, I think we should also change mentions of `Consensus Commit transaction manager` to `Consensus Commit transaction protocol` throughout this doc.

Comment (feeblefakie) [reply docs/configurations.md]: Yeah, I think the title should be revised. I also feel the overall structure might need to be reorganized since the second level titles are a bit inconsistent. (Transaction Managers, ScalarDB Server configurations, and Other ScalarDB configurations)

How about this? > @josh-wong @brfrn169 

# ScalarDB Configurations
## ScalarDB client configurations
### Use Consensus Commit directly
### Use Consensus Commit through ScalarDB Server
## ScalarDB Server configurations
## Other ScalarDB configurations

**Review by Torch3333 — APPROVED** [review]
Besides the comment below, LGTM thank you!

File: docs/configurations.md [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+| `scalar.db.password`                  | AWS secret access key used to authenticate the user interacting with AWS.                                                                                                                                                                                   |            |
+| `scalar.db.dynamo.endpoint_override`  | Amazon DynamoDB endpoint with which ScalarDB should communicate. This is primarily used for testing with a local instance instead of an AWS service.                                                                                                        |            |
+| `scalar.db.dynamo.metadata.namespace` | Namespace name for the table metadata used for ScalarDB.                                                                                                                                                                                      | `scalardb` |
```
[the remark this answers is not in the document]
Comment (josh-wong) [reply docs/configurations.md:101]: No worries. Thanks for letting me know!

File: docs/configurations.md (hunk 17 above)
Comment (josh-wong) [reply docs/configurations.md]: @feeblefakie That structure makes sense to me. I've made revised the headings in https://github.com/scalar-labs/scalardb/commit/f2cf495154bdd0ed4a2fba47c35fd8cd1fe4d117.

**Review by brfrn169 — APPROVED** [review]
LGTM! Thank you!

File: docs/configurations.md [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+| `scalar.db.grpc.max_inbound_message_size`  | The maximum message size allowed for a single gRPC frame.   | The gRPC default value. |
+| `scalar.db.grpc.max_inbound_metadata_size` | The maximum size of metadata allowed to be received.        | The gRPC default value. |
+
+For details about ScalarDB Server, see [ScalarDB Server](scalardb-server.md).
+
+## ScalarDB Server configurations
+
+[ScalarDB Server](scalardb-server.md) is a standalone server that provides a gRPC interface to ScalarDB. This section explains ScalarDB Server configurations.
+
+In addition to the [configurations when using Consensus Commit through ScalarDB Server](#use-consensus-commit-through-scalardb-server) and [other ScalarDB configurations](#other-scalardb-configurations), the following configurations are available for ScalarDB Server:
```
Comment (feeblefakie) [inline docs/configurations.md:160]:
```suggestion
In addition to the [configurations when using Consensus Commit directly](#use-consensus-commit-directly) and [other ScalarDB configurations](#other-scalardb-configurations), the following configurations are available for ScalarDB Server:
```
Isn't this a mistake since ScalarDB Server is the direct client for databases?

@josh-wong @brfrn169

Comment (brfrn169) [reply docs/configurations.md:160]: Yes, that was a mistake. ScalarDB Server should directly use Consensus Commit. Thank you for pointing it out!

Comment (josh-wong) [reply docs/configurations.md]: Thank you for catching this! I've applied your suggestion.

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!
